### DISCUSSION TITLE
How to generate a random sober pastel color and the opposite bright color

### DISCUSSION THREAD

**Discussion by srini8080** [ImageMagick/ImageMagick#4783] [Development]
![Screenshot from 2022-01-30 18-08-58](https://user-images.githubusercontent.com/12046302/151700104-42745b0b-b8cb-4223-8b60-8d5f5ab50c8f.png)

Centos 8, Imagemagick v7.

I want to generate a random sober pastel color like the ones in the image and the opposite bright intense color of it.

Any suggestions?

Thanks,

**Comment by snibgo** [comment]
Many approaches are possible. "Pastel" usually mean "mixed with white". So I first generate a fully-saturated colour, meaning that at least one RGB channel is 100%. Then I blend it with white. I append the two colours side by side. Windows BAT script:
```
magick ^
  -size 1x1 xc:White ^
  -fx "u*rand()" ^
  -evaluate divide %%[fx:maxima] ^
  ( +clone ^
    xc:White ^
    -compose blend -define compose:args=60 -composite ^
  ) ^
  -scale 100x100 ^
  +append ^
  out.png
```
Sample output:
![out](https://user-images.githubusercontent.com/33812368/151709560-2ec8a205-52e9-4cc6-a59a-f621ef0aad79.png)

**Reply by srini8080** [reply]
Thanks much

**Comment by srini8080** [comment]
convert -size 1x1 xc:White -fx "u*rand()" -evaluate Divide [fx:maxima]   \( +clone xc:White -compose blend -define compose:args=60 -composite \)   -scale 100x100 +append out.png
convert: invalid argument for option `-evaluate': [fx:maxima] @ error/convert.c/ConvertImageCommand/1461.
I am getting this error in 6.9.0 version.
    On Sunday, January 30, 2022, 10:38:30 PM GMT+5:30, snibgo ***@***.***> wrote:  
 
 


Many approaches are possible. "Pastel" usually mean "mixed with white". So I first generate a fully-saturated colour, meaning that at least one RGB channel is 100%. Then I blend it with white. I append the two colours side by side. Windows BAT script:
magick ^
  -size 1x1 xc:White ^
  -fx "u*rand()" ^
  -evaluate divide %%[fx:maxima] ^
  ( +clone ^
    xc:White ^
    -compose blend -define compose:args=60 -composite ^
  ) ^
  -scale 100x100 ^
  +append ^
  out.png

Sample output:


—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by fmw42** [comment]
Your original post said 

> Centos 8, Imagemagick v7.

So why are you  using IM 6 convert????

You need IM 7 (magick, not convert) to do %[fx: ... ] computations in-line. You also left off the %

`magick -size 1x1 xc:White -fx "u*rand()" -evaluate Divide "%[fx:maxima]" \( +clone xc:White -compose blend -define compose:args=60 -composite \)  -scale 100x100 +append out.png`

**Reply by fmw42** [reply]
I think this is the equivalent if you really need to use IM 6 (convert). 

`convert -size 1x1 xc:White -fx "u*rand()" \( +clone -separate -evaluate-sequence max \) +swap -compose divide -composite \( +clone xc:White -compose blend -define compose:args=60 -composite \)  -scale 100x100 +append out2.png`

**Comment by srini8080** [comment]
great!.
If I have to know the hex color codes of the generated 2 parts, how do I get that?
    On Monday, January 31, 2022, 10:30:45 AM GMT+5:30, Fred Weinhaus ***@***.***> wrote:  
 
 


I think this is the equivalent if you really need to use IM 6 (convert).

convert -size 1x1 xc:White -fx "u*rand()" \( +clone -separate -evaluate-sequence max \) +swap -compose divide -composite \( +clone xc:White -compose blend -define compose:args=60 -composite \) -scale 100x100 +append out2.png

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by snibgo** [comment]
Before 2-scale", insert "+write txt".

Or insert "-format "%[hex:s]\n" +write info:".

**Comment by srini8080** [comment]
![out028](https://user-images.githubusercontent.com/12046302/151835995-7bc7118b-fb64-4f33-9580-2a51793720f3.png)
![out022](https://user-images.githubusercontent.com/12046302/151836005-848b917b-27d4-4fde-ac84-a9d274d77db0.png)
![out017](https://user-images.githubusercontent.com/12046302/151836008-6b53aa7e-56fe-4b4c-a413-7a023b8b08a5.png)
![out010](https://user-images.githubusercontent.com/12046302/151836011-00f535f4-d332-4d34-bd8c-3dd9d5263491.png)

I am getting some sober colors still instead of bright intense colors.

**Reply by srini8080** [reply]
I mean on the left.....

**Comment by snibgo** [comment]
Define "bright intense colors".

**Comment by srini8080** [comment]
![out049](https://user-images.githubusercontent.com/12046302/151838756-b36100c2-13a7-4c83-942d-6d71f10498e7.png)
![out045](https://user-images.githubusercontent.com/12046302/151838764-890caaff-63b7-4eb7-881d-409aec169572.png)
![out042](https://user-images.githubusercontent.com/12046302/151838766-cb85adbb-2f95-4511-a572-44923d859114.png)
![out036](https://user-images.githubusercontent.com/12046302/151838770-a2972b02-6297-47b4-8e51-326e8235a7df.png)
![out032](https://user-images.githubusercontent.com/12046302/151838775-ea60dd1a-d76d-4d35-bf47-785166195afd.png)
![out030](https://user-images.githubusercontent.com/12046302/151838777-6b49c214-7d51-4ab9-b172-402b90373d82.png)
![out028](https://user-images.githubusercontent.com/12046302/151838780-af914272-c771-4eeb-97fc-7d27f70f6cea.png)

**Reply by srini8080** [reply]
on the left.. bright colors.....

**Reply by fmw42** [reply]
I do not think you understand.  The command appends one bright color with its pastel version.  The left image is bright because one channel is at 100% full intensity. If you do not want the left image, then delete it. But you need to start with it to get the pastel one.  Snibgo, correct me if I am wrong here.

**Comment by snibgo** [comment]
You show some examples, but not a definition of "bright intense colors".

**Reply by srini8080** [reply]
Hmmm.. Tough one to define. Dark colors??

https://imagemagick.org/script/color.php

Like Crimson, Firebrick, Brown ....

**Comment by snibgo** [comment]
If we can define it, we can create it. I used a tentative definition that "at least one RGB channel is 100%". We can add a further condition that "at least one RGB channel is 0%".

```
magick ^
  -size 1x1 xc:White ^
  -fx "u*rand()" ^
  -evaluate subtract %%[fx:minima*QuantumRange] ^
  -evaluate divide %%[fx:maxima] ^
  ( +clone ^
    xc:White ^
    -compose blend -define compose:args=60 -composite ^
  ) ^
  -depth 8 ^
  +write txt: ^
  -format "%%[hex:s]\n" +write info: ^
  -format "%%[pixel:s]\n" +write info: ^
  -scale 100x100 ^
  +append ^
  out.png
```
Or we could work in a different colorspace, such as HCL, where C is 100%, H is random, and L is random or a fixed 50% or a darker lightness or whatever you want.

**Reply by srini8080** [reply]
This worked perfect.

What would be the command for HCL space?

Thanks

**Comment by snibgo** [comment]
A version using HCL is:
```
magick ^
  -size 1x1 xc:#ff8 ^
  -channel 0 -evaluate Multiply %%[fx:rand()] +channel ^
  -set colorspace HCL ^
  -colorspace sRGB ^
  ( +clone ^
    xc:White ^
    -compose blend -define compose:args=60 -composite ^
  ) ^
  -scale 100x100 ^
  +append ^
  x.png
```

**Reply by srini8080** [reply]
Thanks much... It worked well.

Also I upgraded to v7.
